precompiles/bank/method_deposit.go

```diff
@@ -0,0 +1,176 @@
+package bank
+
+import (
+	"math/big"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/zeta-chain/protocol-contracts/v2/pkg/zrc20.sol"
+
+	ptypes "github.com/zeta-chain/node/precompiles/types"
+	"github.com/zeta-chain/node/x/fungible/types"
+)
+
+func (c *Contract) deposit(
+	ctx sdk.Context,
+	method *abi.Method,
+	caller common.Address,
+	args []interface{},
+) (result []byte, err error) {
+	// This function is developed using the
+	// Check - Effects - Interactions pattern:
+	// 1. Check everything is correct.
+	if len(args) != 2 {
+		return nil, &(ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		})
+	}
+
+	// function deposit(address zrc20, uint256 amount) external returns (bool success);
+	ZRC20Addr, amount := args[0].(common.Address), args[1].(*big.Int)
+	if amount.Sign() < 0 || amount == nil || amount == new(big.Int) {
+		return nil, &ptypes.ErrInvalidAmount{
+			Got: amount.String(),
+		}
+	}
+
+	// Initialize the ZRC20 ABI, as we need to call the balanceOf and allowance methods.
+	ZRC20ABI, err := zrc20.ZRC20MetaData.GetAbi()
+	if err != nil {
+		return nil, &ptypes.ErrUnexpected{
+			When: "ZRC20MetaData.GetAbi",
+			Got:  err.Error(),
+		}
+	}
+
+	// Check for enough balance.
+	// function balanceOf(address account) public view virtual override returns (uint256)
+	argsBalanceOf := []interface{}{caller}
+
+	resBalanceOf, err := c.CallContract(ctx, ZRC20ABI, ZRC20Addr, "balanceOf", argsBalanceOf)
```

Comment:
This is the balanceOf for the ERC20 zrc20 balance in the contract address?

This checks is valid but we should think about going further: this is an invariant that should never occur, if it occur it mean something wrong is with the blockchain like ability to mint the ZRC20 in the module

In the future we should look into adding an invariant: https://docs.cosmos.network/v0.46/building-modules/invariants.html

erc20Balance >= minted ZRC20 Cosmos coin

Reply:
Exactly. We check the caller has enough balance of ZRC20 to spend, and then we check the bank has an allowance enough to handle the amount of tokens the caller wants to mint.

Taking a look at invariants.

Reply:
Opened https://github.com/zeta-chain/node/issues/2903

As this would require multiple changes I'd like for this PR to be focused on the bank precompile.